pennylane/beta/tapes/tape.py

```diff
@@ -833,3 +835,105 @@ def data(self):
     @data.setter
     def data(self, params):
         self.set_parameters(params, trainable_only=False)
+
+    # ========================================================
+    # execution methods
+    # ========================================================
+
+    def execute(self, device, params=None):
+        """Execute the tape on a quantum device.
+
+        Args:
+            device (~.Device): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameters are used (via :meth:`~.get_parameters`).
+
+        **Example**
+
+        .. code-block:: python
+
+            from pennylane.beta.tapes import QuantumTape
+            from pennylane.beta.queuing import expval, var, sample, probs
+
+            with QuantumTape() as tape:
+                qml.RX(0.432, wires=0)
+                qml.RY(0.543, wires=0)
+                qml.CNOT(wires=[0, 'a'])
+                qml.RX(0.133, wires='a')
+                probs(wires=[0, 'a'])
+
+        If parameters are not provided, the existing tape parameters are used:
+
+        >>> dev = qml.device("default.qubit", wires=[0, 'a'])
+        >>> tape.execute(dev)
+        array([[8.84828969e-01, 3.92449987e-03, 4.91235209e-04, 1.10755296e-01]])
+
+        Parameters can be optionally passed during execution:
+
+        >>> tape.execute(dev, params=[1.0, 0.0, 1.0])
+        array([[0.5931328 , 0.17701835, 0.05283049, 0.17701835]])
+
+        Parameters provided for execution are temporary, and do not affect
+        the tapes' parameters in-place:
+
+        >>> tape.get_parameters()
+        [0.432, 0.543, 0.133]
+        """
+        if params is None:
+            params = self.get_parameters()
+
+        return self._execute(params, device=device)
+
+    def execute_device(self, params, device):
+        """Execute the tape on a quantum device.
+
+        This is a low-level method, intended to be called by an interface,
+        and does not support autodifferentiation.
+
+        For more details on differentiable tape execution, see :meth:`~.execute`.
+
+        Args:
+            device (~.Device): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameter values are used (via :meth:`~.get_parameters`).
+        """
+        device.reset()
+
+        # backup the current parameters
+        current_parameters = self.get_parameters()
+
+        # temporarily mutate the in-place parameters
+        self.set_parameters(params)
+
+        if isinstance(device, qml.QubitDevice):
+            res = device.execute(self)
+        else:
+            res = device.execute(self.operations, self.observables, {})
+
+        # Update output dim if incorrect.
```

Comment:
How come that in certain cases the output dim needs to be updated? Might be worth noting here perhaps.

Reply:
The output dimension of the QNode needs to be known in order for the Jacobian to be created properly. Currently, the QNode computes the output dimension when processing the queue:

* If a measurement process that returns a scalar is included, the output dim is incremented by 1
* If probability is included, the output dim is incremented by `2**len(wires)`
* If a CV probability is included, the output dim is incremented by `cutoff ** len(wires)`.

However, this is quite inelegent. It means that devices aren't free to define their own output without needing to always make the QNode aware of what the output dimension is (example: when we added CV probability, we needed to update the output dim in the QNode).

Instead, what we can do is add another approach; the tape simply looks out the output size after execution, without needing to know anything in advance!

This has a drawback though --- you always need to evaluate the QNode before computing the Jacobian.